File: day1/in_go/task2.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"strconv"
)

func die(v ...interface{}) {
	fmt.Println(v...)
	os.Exit(1)
}

func main() {
	if len(os.Args) != 2 {
		die("Please specify input text file as only argument")
	}

	file, err := os.Open(os.Args[1])
	if err != nil {
		die(err)
	}
	defer file.Close()

	list := []int{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		val, err := strconv.Atoi(scanner.Text())
		if err != nil {
			die(err)
		}
		list = append(list, val)
	}

	frequency := 0
	seen := map[int]bool{}
	loop := 1

	for {
		for _, diff := range list {
			fmt.Println("Loop:", loop)
			frequency += diff
			_, exists := seen[frequency]

			if exists {
				fmt.Println("Duplicate frequency found:", frequency)
				os.Exit(0)
			} else {
				seen[frequency] = true
			}
			loop++
		}
	}

}

File: day1/in_c/task2.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define LINE_MAX 256
#define ARR_MAX 1234567890

static char line[LINE_MAX];
static int frequency = 0;

static int seen[ARR_MAX];
static int seen_end = 0;

static void
store(int val)
{
	seen[seen_end++] = val;
	
	if (seen_end > ARR_MAX) {
		printf("Make array bigger!\n");
		exit(EXIT_FAILURE);
	}
}

static bool
already_seen(int val)
{
	for (int i = 0; i < seen_end; i++) {
		if (seen[i] == val) {
			return true;
		}
	}
	return false;
}

int
main(int argc, char** argv)
{
	FILE* input;

	if (argc != 2) {
		printf("Please specify input text file as only argument.\n");
		exit(EXIT_FAILURE);
	}

	input = fopen(argv[1], "r");
	if (!input) {
		printf("Not a valid file: %s\n", argv[1]);
		exit(EXIT_FAILURE);
	}

	while (true) {
		while (fgets(line, sizeof(line), input)) {
			frequency += atoi(line);

			if (already_seen(frequency)) {
				printf("Duplicate frequency found: %d\n", frequency);
				exit(EXIT_SUCCESS);
			} else {
				store(frequency);
			}
		}

		fseek(input, 0, SEEK_END);
		fseek(input, 0, SEEK_SET);
	}
}
